Support for DB2

aguibert opened the pull request:
Closes #64

aguibert commented on build.gradle:
```diff
@@ -14,6 +14,9 @@ subprojects {
 
     repositories {
         mavenLocal()
+        maven {
+			url "https://oss.sonatype.org/content/repositories/iovertx-3873"
+		}
```
I will remove this once Vertx 3.9.1 is official released

DavideD commented on README.md:
```diff
@@ -95,12 +95,23 @@ the instructions in [podman.md](podman.md).
+    
+TIP: To reuse the same containers across multiple runs, set `testcontainers.reuse.enable=true` in
+a file located at `$HOME/.testcontainers.properties` (create the file if it does not exist).
+
```
In the future could you keep these changes in a separate commit? These doesn't seem related to DB2. It's more an improvement of the documentation.

DavideD commented on README.md:
```diff
@@ -95,12 +95,23 @@ the instructions in [podman.md](podman.md).
+#### Starting databases automatically with Testcontainers
+
+If you have Docker installed, the container instances will be automatically started
+by running the tests if you select the `-Pdocker` profile like so:
```
I think it's more consistent with the previous paragraph (and shorter) if you you rewrite it to something like: 
```
#### If you have Docker

The build will start the container automatically if you select the profile `-Pdocker`:
```

DavideD replied:
I guess the profile is `docker`  (and not `-Pdocker`).

DavideD commented on README.md:
```diff
@@ -95,12 +95,23 @@ the instructions in [podman.md](podman.md).
+    ./gradlew test -Pdocker
+    
+TIP: To reuse the same containers across multiple runs, set `testcontainers.reuse.enable=true` in
+a file located at `$HOME/.testcontainers.properties` (create the file if it does not exist).
```
I'm suprised that the file is located in the $HOME. Isn't it possible to set for the current project?

aguibert replied:
Nope, the reusable container feature was intentionally designed to have a 2-stage enablement:
1) The container object needs to declare it is reusable by doing `.withReuse(true)`
2) The system where the test is running needs to be configured for reuse (via the properties file)
Item (2) is a bit tedious, but it was put in place mainly so that reusable containers don't get left around on CI machines, since it's only a feature that should be used in local environments.

DavideD replied:
ok

DavideD commented on hibernate-rx-core/build.gradle:
```diff
@@ -17,17 +17,21 @@ dependencies {
     testImplementation 'io.vertx:vertx-unit:3.8.1'
     testImplementation 'org.postgresql:postgresql:42.2.12'
     testImplementation 'mysql:mysql-connector-java:8.0.19'
+    testImplementation 'com.ibm.db2:jcc:11.1.4.4'
```
We just merged https://github.com/hibernate/hibernate-reactive/pull/118, so I don't think this dependency is needed anymore

aguibert replied:
To determine container readiness they establish a JDBC connection, which will require the driver to be on the classpath. However, I can write a custom readiness check for testcontainers that establishes an RX connection instead.

DavideD commented on hibernate-rx-core/src/main/java/org/hibernate/rx/service/RxConnectionPoolProviderImpl.java:
```diff
@@ -59,10 +62,21 @@ private Pool configurePool(Map configurationValues) {
-		final String database = uri.getPath().substring( 1 );
+		String database = uri.getPath().substring( 1 );
+		if (uri.getScheme().equals("db2") && database.indexOf( ':' ) > 0) {
+			database = database.substring( 0, database.indexOf( ':' ) );
```
Maybe you can add a comment to highlight how the db2 uri differs form the other ones

DavideD commented on hibernate-rx-core/src/main/java/org/hibernate/rx/service/RxConnectionPoolProviderImpl.java:
```diff
@@ -59,10 +62,21 @@ private Pool configurePool(Map configurationValues) {
+			    int queryIndex = uri.getPath().indexOf(':') + 1;
+			    if (queryIndex > 0) {
+			      params = uri.getPath().substring(queryIndex).split(";");
+			    }
```
An example of the db2 uri here would help to understand why you need the extra logic

DavideD commented on hibernate-rx-core/src/test/resources/vertx-default-jul-logging.properties:
```diff
@@ -0,0 +1,7 @@
+java.util.logging.SimpleFormatter.format=[%1$tF %1$tT] %4$s %2$-55.55s %5$s%6$s%n
+
+.level=INFO
+io.vertx.db2client.level=FINEST
```
this is another thing that it would be nice to have in a separate commit

DavideD commented on hibernate-rx-core/src/test/java/org/hibernate/rx/DB2BasicTest.java:
```diff
@@ -0,0 +1,308 @@
+				.thenApply(s -> { context.assertTrue(basik.prePersisted && !basik.postPersisted); return s; } )
+				.thenApply(s -> { context.assertTrue(basik.parent.prePersisted && !basik.parent.postPersisted); return s; } )
+				.thenCompose(s -> s.flush())
+				.thenApply(s -> { context.assertTrue(basik.postPersisted && basik.postPersisted); return s; } )
```
I think this is supposed to be:
`basik.prePersisted && basik.postPersisted`

aguibert commented:
thanks for reviewing @DavideD, I'll address these shortly. In the meantime, can you please have a look at my proposal over on this issue? https://github.com/hibernate/hibernate-reactive/issues/61
Now that we have 3 different databases, I'd like like to get a better strategy for testing with multiple DBs

gavinking commented:
@aguibert don't forget to add code to infer the `Dialect` from the JDBC URL, a task which is currently, unsatisfactorily, handled by a hardcoded switch in `RxJdbcEnvironmentInitiator`

aguibert commented:
thanks for the reminder @gavinking, I hadn't noticed the dialect auto-detection changes go in. Added DB2 detection to the PR now.

@DavideD thanks for reviewing -- I've addressed all of your review comments and the PR should be all fixed up now

aguibert commented:
CI is good now except for `AutoincrementTest` which is failing, but not caused by these changes, so I've raised a separate issue (#125) to track that

Sanne commented:
sorry - lots of conflicts need to be resolved now.

Tangential, we might no longer need the custom maven repository configuration?

gavinking commented:
> sorry - lots of conflicts need to be resolved now.

yeah, sorry about that @aguibert :-/

aguibert commented:
ok, all merged and up to date now

DavideD commented:
Thanks @aguibert!

gavinking commented:
Great!!
